js/pattern.js

```diff
@@ -1011,17 +1011,56 @@ function initPattern(globals){
             }
 
             var faceVert = [];
-            for (var j=0;j<face.length;j++){
-                var vertex = vertices[face[j]];
-                faceVert.push(vertex[0]);
-                faceVert.push(vertex[1]);
-                if (!is2d) faceVert.push(vertex[2]);
+            var triangles = [];
+            if (is2d) {
+                for (var j=0;j<face.length;j++){
+                    var vertex = vertices[face[j]];
+                    faceVert.push(vertex[0]);
+                    faceVert.push(vertex[2]);
+                }
+                triangles = earcut(faceVert, null, 2);
+            } else {
+                // earcut only uses the two first coordinates for triangulation...
+                // as a fix, we try each of the dimension combinations until we get a result
+                for (var j=0; j<3; j++) {
+                    faceVert = [];
+                    for (var k=0;k<face.length;k++){
+                        var vertex = vertices[face[k]];
+                        faceVert.push(vertex[j]);
+                        faceVert.push(vertex[(j + 1) % 3]);
+                        faceVert.push(vertex[(j + 2) % 3]);
+                    }
+                    triangles = earcut(faceVert, null, 3);
+                    // make sure we got *enough* triangle to cover the face
+                    if (triangles.length >= 3 * (face.length - 2)) break;
```

Comment:
Just added this update. I found another problem, here is how to recreate:
- Open OrigamiSimulator Map SVG (Vertex tolerance: 1px)
- Save as FOLD (do not triangulate)
- Load FOLD
On my end, I see exactly two triangles missing (I see this in a previous comment too if you notice)

Here is the FOLD file for reference:
[origamisimulator.fold.zip](https://github.com/amandaghassaei/OrigamiSimulator/files/5612887/origamisimulator.fold.zip)

Debugging it, it seems that again because of earcut using the first two coordinates, for some faces (very few..) we get less triangles than needed to cover the face. In this FOLD, there are two 5 vertices faces that earcut returns only two triangles (one vertex is out of the game). 

My fix is checking that the number of triangles is at least the number of vertices on the face - 2. That is always right, right?

We're getting close... :)

Reply:
Yes, assuming no added vertices (which must be the case for earcut, because it has no mechanism to return additional vertices), the number of triangles should be exactly face - 2, so this seems like a good change.

What do you think about also changing the order that we try the dimension pairs to start with x and z, like we do in 2D?  It shouldn't matter, but for files that were originally flat and imported, this seems like a better starting point, and kind of nice to try the `is2d` algorithm and then continue?  We could just reverse the loop...

Reply:
Sounds good to me. Pushed a commit where I simply reversed the loop and checked with the FOLD and SVG models and it all seems fine